### PR TITLE
Non iid data sampler

### PR DESCRIPTION
**Pull request by IzaakC**
Lijkt te werken, maar er mist nog een mogelijkheid om het globale model te testen. De testdata is even non-iid voor allle clients, en geeft dus geen goed beeld van het globale model.  

edit:
Global model test is toegevoegd, en er is een duidelijk verschill in accuracy tiussen non-idd en iid data te zien

edit 2:
Twee andere sample methodes zijn nu ook toegevoegd. q-sampler (sample met label x wordt aan een client van groep x gegeven met probability q), en dirichlet.

### CODE REVIEW

File: fltk/datasets/cifar10.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 from torchvision import transforms
```
Comment (IzaakC) [inline fltk/datasets/cifar10.py:3]: This file should actually be removed to avoid confusion. Only distributed/cifar10.py is used

File: fltk/util/data_sampler_utils.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+    def __iter__(self) -> Iterator[int]:
+        return iter(self.indices)
+
+    def __len__(self) -> int:
+        logging.info("length of client with rank {} : {}".format(self.rank, len(self.indices)))
+        return len(self.indices)
+
+
+if __name__ == "__main__":
```
Comment (bacox) [inline fltk/util/data_sampler_utils.py:221]: Is this used for testing or something else?

File: fltk/datasets/distributed/cifar10.py [hunk 3]
```diff
@@ -39,48 +40,66 @@ def init_test_dataset(self):
...
+    #     test_loader = DataLoader(test_dataset, batch_size=len(test_dataset), sampler=sampler)
+    #     self.args.set_sampler(sampler)
+
+    #     test_data = self.get_tuple_from_data_loader(test_loader)
+
+    #     self.get_args().get_logger().debug("Finished loading CIFAR10 test data")
+
+    #     return test_data
+
+    def get_sampler(self, dataset):
```
Comment (bacox) [inline fltk/datasets/distributed/cifar10.py:88]: I think it is best if the samples are independent of the datasets. Otherwise we would have to implement it for every new dataset. Suggest to move this function to the config or the Datasampler file and load it from there

File: fltk/federator.py [hunk 4]
```diff
@@ -78,6 +78,14 @@ def __init__(self, client_id_triple, num_epochs = 3, config=None):
         self.create_clients(client_id_triple)
         self.config.init_logger(logging)
 
+        logging.info("Creating test client")
+        copy_sampler = config.data_sampler
+        config.data_sampler = "uniform"
+        self.test_data = Client("test", None, 1, 2, config)
+        self.test_data.init_dataloader()
+        config.data_sampler = copy_sampler
```
Comment (bacox) [inline fltk/federator.py:86]: Do we test on the whole (test) dataset or on only the part allocated based on the rank, as the naive distributed sampler does?

**Review by bacox — COMMENTED** [review]
Nice contribution. I've took the liberty to move the get_sampler function.
I think there is only one question at federator.py, apart from that one everything looks good

File: fltk/datasets/cifar10.py (hunk 1 above)
Comment (bacox) [reply fltk/datasets/cifar10.py]: Done, it's removed

File: fltk/util/data_sampler_utils.py (hunk 2 above)
Comment (IzaakC) [reply fltk/util/data_sampler_utils.py]: Only for testing, will remove this.

File: fltk/federator.py (hunk 4 above)
Comment (IzaakC) [reply fltk/federator.py:86]: The federator tests on the whole test dataset. It is actually based on rank, but the world size for this 'client' is set to 2, so the result is the same. Using a client here made it easy to have a global test dataset.
